web/pandas/pdeps/0013-logical-type-system.md

```diff
@@ -0,0 +1,161 @@
+# PDEP-13: The pandas Logical Type System
+
+- Created: 27 Apr 2024
+- Status: Draft
+- Discussion: [#58141](https://github.com/pandas-dev/pandas/issues/58141)
+- Author: [Will Ayd](https://github.com/willayd),
+- Revision: 1
+
+## Abstract
+
+This PDEP proposes a logical type system for pandas to abstract underlying library differences from end users, clarify the scope of pandas type support, and give pandas developers more flexibility to manage the implementation of types.
+
+## Background
+
+When pandas was originally built, the data types that it exposed were a subset of the NumPy type system. Starting back in version 0.23.0, pandas introduced Extension types, which it also began to use internally as a way of creating arrays instead of exclusively relying upon NumPy. Over the course of the 1.x releases, pandas began using pyarrow for string storage and in version 1.5.0 introduced the high level ``pd.ArrowDtype`` wrapper.
+
+While these new type systems have brought about many great features, they have surfaced three major problems. The first is that we put the onus on users to understand the differences of the physical type implementations. Consider the many ways pandas allows you to create a "string":
+
+```python
+dtype=object
+dtype=str
+dtype="string"
+dtype=pd.StringDtype()
+dtype=pd.StringDtype("pyarrow")
+dtype="string[pyarrow]"
+dtype="string[pyarrow_numpy]"
+dtype=pd.ArrowDtype(pa.string())
+```
+
+Keeping track of all of these iterations and their subtle differences is difficult even for [core maintainers](https://github.com/pandas-dev/pandas/issues/58321).
+
+The second problem is that the conventions for constructing types from a given type backend are inconsistent. Let's review string aliases used to construct certain types:
+
+| logical type | NumPy   | pandas extension | pyarrow                  |
+| int          | "int64" | "Int64"          | "int64[pyarrow]"         |
+| string       | N/A     | "string"         | N/A                      |
+| datetime     | N/A     | "datetime64[us]" | "timestamp[us][pyarrow]" |
+
+"string[pyarrow]" is excluded from the above table because it is misleading; while "int64[pyarrow]" definitely gives you a pyarrow backed string, "string[pyarrow]" gives you a pandas extension array which itself then uses pyarrow, which can introduce behavior differences (see [issue 58321](https://github.com/pandas-dev/pandas/issues/58321)).
+
+If you wanted to try and be more explicit about using pyarrow, you could use the ``pd.ArrowDtype`` wrapper. But this unfortunately exposes gaps when trying to use that pattern across all backends:
+
+| logical type | NumPy    | pandas extension | pyarrow                           |
+| int          | np.int64 | pd.Int64Dtype()  | pd.ArrowDtype(pa.int64())         |
+| string       | N/A      | pd.StringDtype() | pd.ArrowDtype(pa.string())        |
+| datetime     | N/A      | ???              | pd.ArrowDtype(pa.timestamp("us")) |
```

Comment:
np.datetime64?

Reply:
How does that work? 

```
In [38]: ser = pd.Series([], dtype=np.datetime64)
ValueError: The 'datetime64' dtype has no unit. Please pass in 'datetime64[ns]' instead.

In [37]: ser = pd.Series([], dtype=np.datetime64("us"))
ValueError: Error parsing datetime string "us" at position 0
```

Reply:
i see the issue, never mind

Reply:
I think strictly speaking the numpy equivalent is `np.dtype("datetime64[ns]")` here, but then you are essentially back at string parsing (which is not what you wanted to show in this table)

Reply:
Ah thanks @jorisvandenbossche . Looks like I also need to update this table - I thought "datetime64[ns]" would map to a pandas extension type. Did not realize that was a numpy data type